Geometry/Surfaces/QuadSurfaces/cone_class.f90

```diff
@@ -0,0 +1,346 @@
+        str = 'zCone'
+
+      case default
+        str = 'unknown cone'
+
+    end select
+
+  end function myType
+
+  !!
+  !! Initialise cone from a dictionary
+  !!
+  !! See surface_inter for more details
+  !!
+  subroutine init(self, dict)
+    class(cone), intent(inout)               :: self
+    class(dictionary), intent(in)            :: dict
+    integer(shortInt)                        :: id, dir
+    real(defReal), dimension(:), allocatable :: vertex
+    character(nameLen)                       :: name
+    real(defReal)                            :: t
+    character(100), parameter :: Here = 'init (cone_class.f90)'
+
+    ! Get from dictionary
+    call dict % get(id, 'id')
+    call dict % get(t, 'tangent')
+    call dict % get(vertex, 'vertex')
+    call dict % get(name, 'type')
+    call dict % get(dir, 'orientation')
+
+    ! Check origin size
+    if (size(vertex) /= 3) then
+      call fatalError(Here,'Vertex needs to have size 3. Has: '//numToChar(size(vertex)))
+    end if
+
+    ! Build cylinder
+    call self % build(id, name, vertex, t, dir)
+
+  end subroutine init
+
+  !!
+  !! Build cone from components
+  !!
+  !! Args:
+  !!   id [in]   -> Surface ID
+  !!   type [in] -> Cone type {'xCone', 'yCone' or 'zCone'}
+  !!   vertex [in]  -> Cone vertex
+  !!   tangent [in] -> Cone opening angle tangent
+  !!   dir [in]     -> Orientation of the cone
+  !!
+  !! Errors:
+  !!   fatalError if id or tangent are -ve, and if orientation is different from
+  !!   1 or -1
+  !!
+  subroutine build(self, id, type, vertex, tangent, dir)
+    class(cone), intent(inout)              :: self
+    integer(shortInt), intent(in)           :: id
+    character(*), intent(in)                :: type
+    real(defReal), dimension(3), intent(in) :: vertex
+    real(defReal), intent(in)               :: tangent
+    integer(shortInt), intent(in)           :: dir
+    character(100), parameter :: Here = 'build (cylinder_class.f90)'
+
+    ! Check values
+   if (id < 1) then
+     call fatalError(Here,'Invalid surface id provided. ID must be > 1')
+
+   else if ( tangent <= ZERO) then
+     call fatalError(Here, 'Tangent of cone must be +ve. It is: '//numToChar(tangent))
+
+   else if (dir /= 1 .and. dir /= -1) then
+     call fatalError(Here, 'Orientation of cone must be 1 or -1. It is: '//numToChar(dir))
+
+   end if
+
+   ! Select type of cylinder
+   select case(type)
+     case('xCone')
+       self % axis = X_AXIS
+       self % plane = [Y_AXIS, Z_AXIS]
+
+     case('yCone')
+       self % axis = Y_AXIS
+       self % plane = [X_AXIS, Z_AXIS]
+
+     case('zCone')
+       self % axis = Z_AXIS
+       self % plane = [X_AXIS, Y_AXIS]
+
+     case default
+       call fatalError(Here, 'Unknown type of cone: '//type)
+
+   end select
+
+   ! Load data
+   self % t_sq   = tangent * tangent
+   self % vertex = vertex
+   self % dir = dir
+   call self % setID(id)
+
+   ! Set surface tolerance
+   call self % setTol(TWO * SURF_TOL)
```

Comment:
It probably should be made consistent with the value in the doc-comment. 

Sadly there is a general problem with the surfaces and surface tolerance I missed when the current geometry was first made. 

I wanted the surface tolerance to be well defined and be approximately a shell of constant 'fuzzy thickness' around the surface, but it sadly will not work with general quadratic surfaces where the 'radius' changes. Hence one cannot get away with the approximation of `2 * r * SURF_TOL` which worked well for cylinders and spheres...   

I think it is fine to leave it as it is but perhaps there should be a comment that will say that 'the thickness of the fuzzy region around the surface will depend on the position on the surface and is not constant'. 

We probably should open an issue about it as well. We need to rethink how the surface tolerance is done.

Reply:
The tests were a bit challenges because of this weird tolerance constant through a changing radius. Now everything should work but I agree that it's a bit annoying and suboptimal.

Hopefully this could be discussed in an issue and addressed in another PR!

Reply:
I changed the tolerance to be proportional to the 'mean radius' of the cone now. That is the radius of the cone at mid-height.